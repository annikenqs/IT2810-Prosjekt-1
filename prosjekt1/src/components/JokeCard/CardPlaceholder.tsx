import "./JokeCard.css";

function CardPlaceHolder() {
	return (
		<section className="card" role="figure">
			
		</section>
	  );
}

export default CardPlaceHolder;
