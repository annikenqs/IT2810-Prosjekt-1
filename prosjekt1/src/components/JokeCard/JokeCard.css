.card {
    position: relative;
    background-color: #dcfbe3;
    font-style: oblique;
    min-height: 200px;
    width: 350px;
    border-radius: 20px;
    box-shadow: 0 10px 15px -3px rgb(0 0 0 / 0.2), 0 4px 6px -4px rgb(0 0 0 / 0.2);
    font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
    padding: 15px;
    text-align: center;
}

@media (max-width: 430px) {
  .card {
    width: 95%;
    padding: 11px;
    min-height: 180px;
  }
}

.favorite-button {
    position: absolute;
    top: 5px;
    right: 5px;
    background: none;
    border: none;
    font-size: 1.5rem;
    cursor: pointer;
}
  
.favorite-button:focus {
    outline: none;
}