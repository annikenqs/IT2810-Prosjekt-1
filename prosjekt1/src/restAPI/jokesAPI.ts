import { useQuery} from '@tanstack/react-query';

export interface JokeResponse {
	type: string;
	joke?: string; // For single-type jokes
	setup?: string; // For two-part jokes
	delivery?: string; // For two-part jokes
	id: number;
	category: string;
}

export const validIDs: number[] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 241, 243, 244, 245, 249, 250, 251, 252, 253, 293, 183, 295, 296, 297, 298, 299, 300, 311, 313, 315];

// Fetch a joke, given a category
const fetchJoke = async (type: string): Promise<JokeResponse> => {
	const response = await fetch(`https://v2.jokeapi.dev/${type}?blacklistFlags=racist,sexist&idRange=0-318`);
	if (!response.ok) {
		throw new Error("Failed to fetch joke");
	} else if (!(type=='Christmas' || type=='Spooky' || type=='Programming')) {
		throw new Error("Invalid joke category");
	}
	const data: JokeResponse = await response.json();
	return data;
}

//Fetch a joke, given an ID (the ID has to be valid)
const fetchJokeById = async (id: number): Promise<JokeResponse> => {
	if (!validIDs.includes(id)) {
		throw new Error("Invalid joke ID");
	}
	const response = await fetch(`https://v2.jokeapi.dev/joke/Programming,Spooky,Christmas?blacklistFlags=racist,sexist&idRange=${id}`);
	if (!response.ok) {
		throw new Error("Failed to fetch joke");
	}
	const data: JokeResponse = await response.json();
	return data;
}

//Fetch jokes for a specific category
const fetchJokesByCategory = async (category: string, batchSize: number, rangeStart: number, rangeEnd: number): Promise<JokeResponse[]> => {
  const response = await fetch(
    `https://v2.jokeapi.dev/joke/${category}?blacklistFlags=racist,sexist&amount=${batchSize}&idRange=${rangeStart}-${rangeEnd}`
  );
  if (!response.ok) {
    throw new Error(`Failed to fetch ${category} jokes`);
  }
  const data = await response.json();
  if (data.error || (data.jokes && data.jokes.length === 0)) {
    console.log(`No jokes found for ${category}. Error:`, data.error ? data.error : 'None found');
    return [];
  }
  return data.jokes ?? [];
};

// Fetch exactly 10 jokes from each category
const fetchAllJokes = async (): Promise<JokeResponse[]> => {
  const batchSize = 10;
  let allJokes: JokeResponse[] = [];
	//Christmas IDs: 241, 243 - 245, 249, 250 - 253, 293
  const christmasJokes = await fetchJokesByCategory("Christmas", batchSize, 241, 293);
  if (christmasJokes.length == 10) allJokes = [...allJokes, ...christmasJokes];
	//Spooky IDs: 183, 295 - 300, 311, 313, 315
  const spookyJokes: JokeResponse[] = await fetchJokesByCategory("Spooky", batchSize, 183, 315);
  if (spookyJokes.length == 10) allJokes = [...allJokes, ...spookyJokes];
	//Programming IDs: 0-9
  const programmingJokes = await fetchJokesByCategory("Programming", batchSize, 0, 10);
  if (programmingJokes.length == 10) allJokes = [...allJokes, ...programmingJokes];
  return allJokes;
};

// Get a random joke
export const useRandomJoke = () => {
	return useQuery({
		queryKey:["joke", "Any"],
		 queryFn: () => fetchJoke("Any"),
	})
}

// Get a programming joke
export const useProgrammingJoke = () => {

	return useQuery({
		queryKey: ["joke", "Programming"],
		queryFn: () => fetchJoke("Programming"),
	});
};

// Get a spooky joke
export const useSpookyJoke = () => {

	return useQuery({
		queryKey: ["joke", "Spooky"],
		queryFn: () => fetchJoke("Spooky"),
	});
};

// Get a christmas joke
export const useChristmasJoke = () => {
	
	return useQuery({
		queryKey: ["joke", "Christmas"],
		queryFn: () => fetchJoke("Christmas"),
	});
};


// Get a joke by ID
export const useJokeById = (id: number) => {
	return useQuery({
		queryKey: ["joke", id],
		queryFn: () => fetchJokeById(id),
		refetchInterval: 1000 * 60 * 10, // Refetch every 10 min
	});
};

//Get all relevant jokes
export const useAllJokes = () => {
  return useQuery({
    queryKey: ["joke", "All"],
    queryFn: () => fetchAllJokes(),
		refetchInterval: 1000 * 60 * 10,
  });
};