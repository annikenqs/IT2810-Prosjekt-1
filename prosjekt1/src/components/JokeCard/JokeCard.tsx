import { useEffect, useState } from "react";
import "./JokeCard.css";
import { inUseIDs, JokeResponse } from "../../restAPI/jokesAPI";

interface JokeCardProps {
	jokeResponse: JokeResponse;
}

function JokeCard({ jokeResponse: j }: JokeCardProps) {
	const [joke, setJoke] = useState<string>("Loading...");
	const [isFavorite, setIsFavorite] = useState<boolean>(false);

	useEffect(() => {
		if (j.isLoading) {
			setJoke("Loading...");
		} else if (j.error) {
			setJoke(j.error.message);
		} else if (j) {
			if (j.type === "single" && j.joke) {
				setJoke(j.joke);
			} else if (j.type === "twopart" && j.setup && j.delivery) {
				setJoke(`${j.setup} - ${j.delivery}`);
			}
		}
	}, [j]);

	useEffect(() => {
		// Load favorite status from localStorage
		const savedFavorites = JSON.parse(localStorage.getItem("favorites") || "[]");
		const isCurrentJokeFavorite = savedFavorites.includes(joke);
		setIsFavorite(isCurrentJokeFavorite);
	}, [joke]);

	const handleFavoriteClick = () => {
		const savedFavorites = JSON.parse(localStorage.getItem("favorites") || "[]");
		if (isFavorite) {
			// Remove the joke from favorites
			const updatedFavorites = savedFavorites.filter((favJoke: string) => favJoke !== joke);
			localStorage.setItem("favorites", JSON.stringify(updatedFavorites));
		} else {
			// Add the joke to favorites
			const updatedFavorites = [...savedFavorites, joke];
			localStorage.setItem("favorites", JSON.stringify(updatedFavorites));
		}
		setIsFavorite(!isFavorite);
	};

	return (
		<section className="card" role="figure">
			<h3>Joke #{inUseIDs.indexOf(j.id) + 1}</h3>
			<p>- {joke}</p>
			{/* Star: to favorite a joke */}
			<button onClick={handleFavoriteClick} className="favorite-button">
				{isFavorite ? "★" : "☆"}
			</button>
		</section>
	  );
}

export default JokeCard;
