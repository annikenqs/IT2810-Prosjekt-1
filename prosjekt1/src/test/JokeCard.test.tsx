import React from "react";
import JokeCard from "../components/JokeCard/JokeCard";
import { JokeResponse } from "../restAPI/jokesAPI";
import { beforeEach, describe } from "vitest";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";

// Define the test suite for the JokeCard component
describe("JokeCard", () => {
	// Mock implementation of localStorage for testing purposes
	const mockLocalStorage = (() => {
		let store: Record<string, string> = {};
		return {
			getItem(key: string) {
				return store[key] || null;
			},
			setItem(key: string, value: string) {
				store[key] = value.toString();
			},
			clear() {
				store = {};
			},
			removeItem(key: string) {
				delete store[key];
			},
		};
	})();
	Object.defineProperty(window, "localStorage", { value: mockLocalStorage });

	// Mock data for a single joke
	const mockSingleJokeResponse: JokeResponse = {
		id: 0,
		type: "single",
		joke: "I've got a really good UDP joke to tell you but I don’t know if you'll get it.",
		isLoading: false,
		error: null,
		category: "Programming",
	};

	// Mock data for a two-part joke (setup and punchline)
	const mockTwopartJokeResponse: JokeResponse = {
		id: 6,
		type: "twopart",
		setup: "Why did the web developer walk out of a resturant in disgust?",
		delivery: "The seating was laid out in tables.",
		isLoading: false,
		error: null,
		category: "Programming",
	};

	// Mock data to simulate a loading state
	const mockLoadingJokeResponse: JokeResponse = {
		id: 0,
		type: "single",
		joke: "I've got a really good UDP joke to tell you but I don’t know if you'll get it.",
		isLoading: true,
		error: null,
		category: "Programming",
	};

	// Mock data to simulate an error state
	const mockErrorJokeResponse: JokeResponse = {
		id: 0,
		type: "single",
		joke: "I've got a really good UDP joke to tell you but I don’t know if you'll get it.",
		isLoading: false,
		error: { name: "Error", message: "Failed to load joke" },
		category: "Programming",
	};

	// Clear localStorage before each test to ensure a clean slate
	beforeEach(() => {
		localStorage.clear();
	});

	// Clean up the rendered components after each test to avoid memory leaks
	afterEach(() => {
		cleanup();
	});

	// Test case 1: Render a single joke and match the output with the stored snapshot
	test("renders JokeCard correctly and matches snapshot for single joke", () => {
		const { container } = render(<JokeCard jokeResponse={mockSingleJokeResponse} />);
		expect(container).toMatchSnapshot();
	});

	// Test case 2: Render a two-part joke and match the output with the stored snapshot
	test("renders JokeCard correctly and matches snapshot for twopart joke", () => {
		const { container } = render(<JokeCard jokeResponse={mockTwopartJokeResponse} />);
		expect(container).toMatchSnapshot();
	});

	test("applies slideshow-card class when isSlideshowCard is true", () => {
		render(<JokeCard jokeResponse={mockSingleJokeResponse} isSlideshowCard={true} />);
		const card = screen.getByRole("figure");
		expect(card).toHaveClass("slideshow-card");
	});

	test("applies slideshow-card class when isSlideshowCard is true", () => {
		render(<JokeCard jokeResponse={mockSingleJokeResponse} isSlideshowCard={true} />);
		const card = screen.getByRole("figure");
		expect(card).toHaveClass("slideshow-card");
	});

	// Test case 3: Display a single joke along with its joke number
	test("displays single joke and joke number", () => {
		render(<JokeCard jokeResponse={mockSingleJokeResponse} />);
		expect(
			screen.getByText(
				/I've got a really good UDP joke to tell you but I don’t know if you'll get it./,
			),
		).toBeInTheDocument();
		expect(screen.getByText(/Joke #1/)).toBeInTheDocument();
	});

	// Test case 4: Display a two-part joke and its joke number
	test("displays single joke and joke number", () => {
		render(<JokeCard jokeResponse={mockTwopartJokeResponse} />);
		expect(
			screen.getByText(/Why did the web developer walk out of a resturant in disgust?/),
		).toBeInTheDocument();
		expect(screen.getByText(/The seating was laid out in tables./)).toBeInTheDocument();
		expect(screen.getByText(/Joke #7/)).toBeInTheDocument();
	});

	// Test case 5: Display "Loading" when the joke is in a loading state
	test("displays loading when loading is true", () => {
		render(<JokeCard jokeResponse={mockLoadingJokeResponse} />);
		expect(screen.getByText(/Loading/)).toBeInTheDocument();
	});

	// Test case 6: Display an error message if there's a failure in loading the joke
	test("displays error message", () => {
		render(<JokeCard jokeResponse={mockErrorJokeResponse} />);
		expect(screen.getByText(/Failed to load joke/)).toBeInTheDocument();
	});

	// Test case 7: Test the favoriting and unfavoriting functionality
	test("handles favoriting and unfavoriting correctly", () => {
		render(<JokeCard jokeResponse={mockSingleJokeResponse} />);

		//click favorite joke
		const favoriteButton = screen.getByRole("button", { name: "☆" });
		fireEvent.click(favoriteButton);
		expect(localStorage.getItem("favorites")).toContain(mockSingleJokeResponse.joke);

		//check if the star button gets filled when favorited
		expect(screen.getByRole("button", { name: "★" })).toBeInTheDocument();
		expect(screen.queryByRole("button", { name: "☆" })).toBeNull();

		//click unfavorite joke
		const favoriteButtonAfter = screen.getByRole("button", { name: "★" });
		fireEvent.click(favoriteButtonAfter);
		expect(screen.getByRole("button", { name: "☆" })).toBeInTheDocument();

		//check if the star button gets unfilled when unfavorited
		expect(screen.queryByRole("button", { name: "★" })).toBeNull();
		expect(localStorage.getItem("favorites")).not.toContain(mockSingleJokeResponse.joke);
	});
});
