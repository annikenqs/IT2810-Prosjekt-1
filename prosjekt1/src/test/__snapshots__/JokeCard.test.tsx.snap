// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JokeCard > renders JokeCard correctly and matches snapshot for single joke 1`] = `
<div>
  <section
    class="card"
    role="figure"
  >
    <h3>
      Joke #
      1
    </h3>
    <p>
      - 
      I've got a really good UDP joke to tell you but I don’t know if you'll get it.
    </p>
    <button
      class="favorite-button"
    >
      ☆
    </button>
  </section>
</div>
`;

exports[`JokeCard > renders JokeCard correctly and matches snapshot for twopart joke 1`] = `
<div>
  <section
    class="card"
    role="figure"
  >
    <h3>
      Joke #
      7
    </h3>
    <p>
      - 
      Why did the web developer walk out of a resturant in disgust? - The seating was laid out in tables.
    </p>
    <button
      class="favorite-button"
    >
      ☆
    </button>
  </section>
</div>
`;
